Refactor trusted_hosts handling

lazyfrosch opened the pull request:
On HTTPS we will get the FQDN from the client certificate and check against the
list.

While on HTTP we will perform both reverse DNS and forward DNS lookup to verify
the client may talk to us.

Refs #7849

lazyfrosch commented:
Tested in my environment, with both HTTPS from an IP that was no reverse lookup, and connection from localhost via trusted_hosts and ip.

domcleal commented:
[test] although we'll probably have to update the trusted hosts test.

lazyfrosch commented:
Are you using a patched version of "test-unit"?

I can't find a method "sorted" nowhere for that gem?

My test fails like this:

```
.../test/unit/test-suite-creator.rb:51:in `collect_test_names': undefined method `sort_test_names_in_sorted_order' for #<Test::Unit::TestSuiteCreator:0x00000003de8d68> (NoMethodError
```

domcleal commented:
No, I've not seen that before!

dmitri-d commented:
Looks good, a couple of notes:
- could you extract branches of the condition at https://github.com/theforeman/smart-proxy/pull/246/files#diff-d00662426e39c5a083182f33fc43ccb2R19 into methods? A bit hard to read as it is atm..
  - could you add some tests for both http and https (let me know if you need help with this)

lzap commented on lib/sinatra/trusted_hosts.rb:
```diff
@@ -6,17 +7,59 @@ def authorize_with_trusted_hosts
           if !Proxy::SETTINGS.trusted_hosts.include?(remote_fqdn.downcase)
             log_halt 403, "Untrusted client #{remote_fqdn.downcase} attempted to access #{request.path_info}. Check :trusted_hosts: in settings.yml"
           end
+
```
Why do we just throw the `remote_fqdn` out on line 37. I understand that client can send any certificate he wants with arbitrary CN, but at least for correct queries both CA name and reverse lookup should both match.

lazyfrosch replied:
Reverse Lookup is not necessarily valid at all times, e.g. I have a VPN between all my root servers, there is no valid reverse-forward lookup.

And also in other areas like dynamic ip addressing might reverse lookup be more or less a nice to have.

The certificate is still validated if it is signed by the Puppet CA.

dmitri-d commented:
Re: failing tests (also see https://github.com/sinatra/sinatra/issues/951) : Looks like we either need to temporary pin rack to `~> 1.5.0`, or use sinatra from master:HEAD. @domcleal -- I think the former is a bit easier, considering that I don't know how temporary "temporary" is.

domcleal commented:
Can we pin to < 1.6?  We have some older versions in circulation...

dmitri-d commented:
Yeah, I think so, will check (and do a PR).

dmitri-d commented:
https://github.com/theforeman/smart-proxy/pull/249 should fix tests failures.

lazyfrosch commented:
Yeah, now actual errors. Will do an update these days.

dmitri-d commented:
[test]

lazyfrosch commented:
So, I've moved the actual X509 and Resolv stuff into the helpers, we now have a smaller TrustedHosts class.

I've added tests to validate resolving and cert handling.

The forward verify can also be disabled by config, in special environments where this is not possible.

Please review.

lazyfrosch committed:
Fixes #7849 - re-factor trusted_hosts handling

On HTTPS we will get the FQDN from the client certificate and check against the
list.

While on HTTP we will perform both reverse DNS and forward DNS lookup to verify
the client may talk to us.

Additionally the forward_verify of DNS can be disabled.

domcleal commented:
[test]

domcleal commented:
[test] some intermittent GitHub errors.

lazyfrosch commented:
I addressed the rubocop warnings...

[test] - nice try, I know :D

dmitri-d commented on test/sinatra/trusted_hosts_test.rb:
```diff
@@ -28,27 +28,69 @@ def test_trusted_hosts_empty
+  def test_trusted_hosts_https_matches
+    Proxy::SETTINGS.expects(:trusted_hosts).at_least_once.returns(['host.example.org'])
+    OpenSSL::X509::Certificate.stubs(:new).returns(OpenSSL::X509::Certificate)
+    OpenSSL::X509::Certificate.expects(:subject).twice.returns('CN=host.example.org,...')
```
you can probably relax the number of calls constraint to `at_least_once` here.

dmitri-d commented:
[test]

dmitri-d commented:
[test]

dmitri-d commented:
lgtm. Will test this in a bit. @domcleal: ping?

domcleal commented on lib/proxy/helpers.rb:
```diff
@@ -19,4 +22,51 @@ def log_halt code=nil, exception=nil
+        log_halt 403, 'could not read CN from the client certificate'
+      end
+    rescue OpenSSL::X509::CertificateError => e
+        log_halt 403, "could not parse the client certificate\n\n#{e.message}"
```
Nitpick, some additional whitespace here.

domcleal commented:
:+1:  looks great!

dmitri-d commented:
tested, lgtm.

dmitri-d commented:
merged in at 30aff66. Thanks, @lazyfrosch!

domcleal commented:
Thanks @lazyfrosch, this will be a great help for users without rDNS set up!

lazyfrosch commented:
@domcleal glad you like it :)
